provider/cluster/service_transfer_hostnames.go

```diff
@@ -0,0 +1,286 @@
+package cluster
+
+import (
+	"context"
+	"errors"
+	"github.com/ovrclk/akash/provider/event"
+	"github.com/ovrclk/akash/pubsub"
+	dtypes "github.com/ovrclk/akash/x/deployment/types"
+	mtypes "github.com/ovrclk/akash/x/market/types"
+	"github.com/tendermint/tendermint/libs/log"
+	"time"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+)
+
+type transferHostnamesRequest struct {
+	hostnames []string
+	destination dtypes.DeploymentID
+	errCh chan <- error
+}
+
+type transferHostnamesManager struct {
+	hostnames []string
+	ownerAddr sdktypes.Address
+	destinationID dtypes.DeploymentID
+	toPurge map[mtypes.LeaseID]*deploymentManager
+	destination []*deploymentManager
+
+	cancel context.CancelFunc
+
+	log log.Logger
+
+	client Client
+}
+
+var errDestintationDeploymentDoesNotExist = errors.New("destination deployment does not exist")
+var errTransferRunning = errors.New("hostname transfer already running")
+
+func (s *service) transferHostnames(req transferHostnamesRequest) error {
+	ownerAddr, err := req.destination.GetOwnerAddress()
+	if err != nil {
+		return err
+	}
+
+	// One transfer running at any time per owner, to avoid any strange race conditions
+	_, exists := s.xferHostnameManagers[ownerAddr.String()]
+	if exists {
+		return errTransferRunning
+	}
+
+	xferMgr := &transferHostnamesManager{
+		hostnames: req.hostnames,
+		ownerAddr: ownerAddr,
+		destinationID: req.destination,
+		toPurge: make(map[mtypes.LeaseID]*deploymentManager),
+		log : s.log.With("owner", ownerAddr.String()),
+		client: s.client,
+	}
+	s.xferHostnameManagers[ownerAddr.String()] = xferMgr
+
+	// Find all deployment managers that need to be forced to update
+	for managerLeaseID, mgr := range s.managers {
+		leaseOwnerAddr, err := managerLeaseID.DeploymentID().GetOwnerAddress()
+		if err != nil {
+			return err
+		}
+		// sort into groups to either be purged or the destination groups
+		if req.destination.Equals(managerLeaseID.DeploymentID()) {
+			xferMgr.destination = append(xferMgr.destination, mgr)
+		} else if ownerAddr.Equals(leaseOwnerAddr) {
+			xferMgr.toPurge[managerLeaseID] = mgr
+		}
+	}
+
+	// Check that the destination is running
+	if len(xferMgr.destination) == 0 {
+		xferMgr.log.Error("destination deployment does not exist", "dseq", req.destination.DSeq)
+		return errDestintationDeploymentDoesNotExist
+	}
+
+	var ctx context.Context
+	ctx, xferMgr.cancel = context.WithCancel(context.Background())
+	go func() {
+		select {
+		case <- s.lc.ShuttingDown():
+			xferMgr.cancel()
+		case <- ctx.Done():
+			return
+		}
+	}()
```

Comment:
can you cancel all `xferHostnameManagers`s when shutting down in the main loop instead of doing this?

Reply:
That should work.